### PR TITLE
ARROW-9241: [C++] Add forward compatibility check for Decimal bit width

### PR DESCRIPTION
**Pull request by wesm**
Also updates the generated Flatbuffers bindings.

### CODE REVIEW

File: cpp/src/arrow/ipc/metadata_internal.cc [hunk 1]
```diff
@@ -257,6 +259,9 @@ Status ConcreteTypeFromFlatbuffer(flatbuf::Type type, const void* type_data,
       return Status::OK();
     case flatbuf::Type::Decimal: {
       auto dec_type = static_cast<const flatbuf::Decimal*>(type_data);
+      if (dec_type->bitWidth() != kDecimalBitWidth) {
+        return Status::Invalid("Library only supports 128-bit decimal values");
```
[the remark this answers is not in the document]
Comment (tianchen92) [reply cpp/src/arrow/ipc/metadata_internal.cc:263]: sorry, bitWidth should return 128 if no value is set. LGTM

**Review by tianchen92 — APPROVED** [review]

File: cpp/src/arrow/ipc/metadata_internal.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (wesm) [reply cpp/src/arrow/ipc/metadata_internal.cc:263]: The Flatbuffers library takes care of this -- `bitWidth()` always returns a value and if the serialized field is not present, the default 128 is returned.

[the remark this answers is not in the document]
Comment (wesm) [reply cpp/src/arrow/ipc/metadata_internal.cc:263]: Here's the generated binding -- so if `VT_BITWIDTH` is not present in the table then it returns 128

```c++
  int32_t bitWidth() const {
    return GetField<int32_t>(VT_BITWIDTH, 128);
  }
```

**Comment by wesm** [comment]
+1

**Pull request closed without merging**